drivers/crypto/crypto_nrf_ecb.c

```diff
@@ -0,0 +1,156 @@
+#include <logging/log.h>
+#include <hal/nrf_ecb.h>
+
+#define ECB_AES_KEY_SIZE   16
+#define ECB_AES_BLOCK_SIZE 16
+
+LOG_MODULE_REGISTER(crypto_nrf_ecb, CONFIG_CRYPTO_LOG_LEVEL);
+
+struct ecb_data {
+	u8_t key[ECB_AES_KEY_SIZE];
+	u8_t cleartext[ECB_AES_BLOCK_SIZE];
+	u8_t ciphertext[ECB_AES_BLOCK_SIZE];
+};
+
+struct nrf_ecb_drv_state {
+	struct ecb_data data;
+	bool in_use;
+};
+
+static struct nrf_ecb_drv_state drv_state;
+
+static int do_ecb_encrypt(struct cipher_ctx *ctx, struct cipher_pkt *pkt)
+{
+	ARG_UNUSED(ctx);
+
+	if (pkt->in_len != ECB_AES_BLOCK_SIZE) {
+		LOG_ERR("unsupported input block size");
+		return -EINVAL;
+	}
+	if (pkt->out_buf_max < pkt->in_len) {
+		LOG_ERR("output buffer too small");
+		return -EINVAL;
+	}
+
+	if (pkt->in_buf != drv_state.data.cleartext) {
+		memcpy(drv_state.data.cleartext, pkt->in_buf, ECB_AES_BLOCK_SIZE);
+	}
+
+	nrf_ecb_event_clear(NRF_ECB, NRF_ECB_EVENT_ENDECB);
+	nrf_ecb_event_clear(NRF_ECB, NRF_ECB_EVENT_ERRORECB);
+	nrf_ecb_task_trigger(NRF_ECB, NRF_ECB_TASK_STARTECB);
+	while (!(nrf_ecb_event_check(NRF_ECB, NRF_ECB_EVENT_ENDECB) ||
+	       nrf_ecb_event_check(NRF_ECB, NRF_ECB_EVENT_ERRORECB))) {
+	}
+	if (nrf_ecb_event_check(NRF_ECB, NRF_ECB_EVENT_ERRORECB)) {
+		LOG_ERR("ECB operation error");
+		return -EIO;
+	}
+	if (pkt->out_buf != drv_state.data.ciphertext) {
+		memcpy(pkt->out_buf, drv_state.data.ciphertext, ECB_AES_BLOCK_SIZE);
+	}
+	pkt->out_len = pkt->in_len;
+	return 0;
+}
+
+static int nrf_ecb_driver_init(struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	nrf_ecb_data_pointer_set(NRF_ECB, &drv_state.data);
+	drv_state.in_use = false;
+	return 0;
+}
+
+static int nrf_ecb_query_caps(struct device *dev)
+{
+	return (CAP_RAW_KEY | CAP_SEPARATE_IO_BUFS | CAP_SYNC_OPS);
+}
+
+static int nrf_ecb_session_setup(struct device *dev, struct cipher_ctx *ctx,
+				 enum cipher_algo algo, enum cipher_mode mode,
+				 enum cipher_op op_type)
+{
+	ARG_UNUSED(dev);
+
+	if (algo != CRYPTO_CIPHER_ALGO_AES) {
+		LOG_ERR("unsupported algorithm");
+		return -EINVAL;
+	}
+
+	if (!(ctx->flags & CAP_SYNC_OPS)) {
+		LOG_ERR("Asynchronous mode not supported by this driver");
+		return -EINVAL;
+	}
+
+	if (ctx->keylen != ECB_AES_KEY_SIZE) {
+		/* driver supports 128 bit key */
+		LOG_ERR("Unsupported key size");
+		return -EINVAL;
+	}
+
+	if (ctx->key.bit_stream == NULL) {
+		LOG_ERR("No key provided");
+		return -EINVAL;
+	}
+
+	if (drv_state.in_use) {
+		LOG_ERR("Peripheral in use");
+		return -EBUSY;
+	} else {
+		drv_state.in_use = true;
+	}
+
+	if (op_type == CRYPTO_CIPHER_OP_ENCRYPT) {
+		switch (mode) {
+		case CRYPTO_CIPHER_MODE_ECB:
+			ctx->ops.block_crypt_hndlr = do_ecb_encrypt;
+			break;
+		default:
+			LOG_ERR("Unsupported mode");
+			drv_state.in_use = false;
+			return -EINVAL;
+		}
+	} else {
+		LOG_ERR("Unsupported operation");
+		drv_state.in_use = false;
+		return -EINVAL;
+	}
+
+	ctx->ops.cipher_mode = mode;
+
+	memcpy(drv_state.data.key, ctx->key.bit_stream, ECB_AES_KEY_SIZE);
```

Comment:
For consistency, you should protect this copying in the same way you did it for packet buffers. I guess the motivation for this was an assumption that a user could use for instance `nrf_ecb_data_pointer_get(NRF_ECB)` to get these pointer values.